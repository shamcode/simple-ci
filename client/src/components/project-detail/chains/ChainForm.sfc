<template>
    {% import CodeFlask from '../../CodeFlask.sfc' %}

    <div class="project-chain-form">
        <form :onsubmit={{this$._submitForm}} :ref="formNode">
            <div class="group">
                <input name="name" type="text" required value={{name}}/>
                <span class="highlight"></span>
                <span class="bar"></span>
                <label>Name</label>
            </div>
            <div class="group">
                <CodeFlask
                    name="command"
                    value={{command}}
                />
                <span class="highlight"></span>
                <span class="bar"></span>
                <label>Command</label>
            </div>
            <div class="group">
                <button type="submit" :disabled={{dataSaving}} >
                    <span>{{saveButtonText}}</span>
                </button>
            </div>
        </form>
    </div>
</template>

<script>
    function extractValue( value ) {
        return value.trim();
    }

    function validateValue( value ) {
        return value.length > 0;
    }

    function ProjectForm( options ) {
        const onSubmit = $();

        const state = options( {
            [ $.name ]: '',
            [ $.command ]: '',
            [ $.saveButtonText ]: '',
            [ $.dataSaving ]: false,
            [ onSubmit ]() {}
        } );

        this$._submitForm = e => {
            e.preventDefault();
            const formData = new FormData( this.formNode );
            const data = {};
            let dataValid = true;
            [
                'name'
            ].forEach( ( key ) => {
                const value = extractValue( formData.get( key ) );
                dataValid = dataValid && validateValue( value );
                if ( dataValid ) {
                    data[ key ] = value;
                }
            } );
            if ( dataValid ) {
                data.command = extractValue( formData.get( 'command' ) );
                state[ onSubmit ]( data );
            }
        };
    }

    export default Component( Template, ProjectForm );
</script>
