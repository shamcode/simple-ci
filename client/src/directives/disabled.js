export default class {
    constructor() {
        this.node = null;
    }

    bind( node ) {
        this.node = node;
    }

    unbind() {
        this.node = null;
    }

    update( value ) {
        this.node.disabled = value;
    }
}
