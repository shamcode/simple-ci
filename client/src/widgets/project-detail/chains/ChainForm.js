import { options } from 'sham-ui';
import ChainFormTemplate from './ChainForm.sht';

function extractValue( value ) {
    return value.trim();
}

function validateValue( value ) {
    return value.length > 0;
}

export default class ProjectForm extends ChainFormTemplate {
    @options get name() { return ''; }
    @options get saveButtonText() { return ''; }
    @options get dataSaving() { return false; }
    @options onSubmit() {}

    get formNode() {
        return this.querySelector( 'form' );
    }

    _submitForm( e ) {
        e.preventDefault();
        const formData = new FormData( this.formNode );
        const data = {};
        let dataValid = true;
        [
            'name',
        ].forEach( ( key ) => {
            const value = extractValue( formData.get( key ) );
            dataValid = dataValid && validateValue( value );
            if ( dataValid ) {
                data[ key ] = value;
            }
        } );
        if ( dataValid ) {
            this.options.onSubmit( data );
        }
    }
}