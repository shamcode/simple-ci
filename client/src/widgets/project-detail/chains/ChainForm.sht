{% import CodeFlask from '../../CodeFlask' %}

<div class="project-chain-form">
    <form :onsubmit={{this._submitForm.bind(this)}}>
        <div class="group">
            <input name="name" type="text" required value={{name}}/>
            <span class="highlight"></span>
            <span class="bar"></span>
            <label>Name</label>
        </div>
        <div class="group">
            <CodeFlask
                name="command"
                value={{command}}
            />
            <span class="highlight"></span>
            <span class="bar"></span>
            <label>Command</label>
        </div>
        <div class="group">
            <button type="submit" :disabled={{dataSaving}} >
                <span>{{saveButtonText}}</span>
            </button>
        </div>
    </form>
</div>
