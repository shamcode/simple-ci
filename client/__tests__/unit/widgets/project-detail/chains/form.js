import renderer from 'sham-ui-test-helpers';
import ChainForm from '../../../../../src/widgets/project-detail/chains/ChainForm';
import directives from '../../../../../src/directives/event-listener';
import disabled from '../../../../../src/directives/disabled';

it( 'renders correctly', () => {
    const meta = renderer( ChainForm, {
        directives: {
            disabled,
            ...directives
        }
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );


it( 'submit form', () => {
    const onSubmitMock = jest.fn();
    const { widget } = renderer( ChainForm, {
        saveButtonText: 'Create',
        directives: {
            disabled,
            ...directives
        },
        onSubmit: onSubmitMock
    } );

    const formData = {
        name: 'Test name',
    };
    widget.querySelector( '[name="name"]' ).value = formData.name;
    widget.querySelector( '[type="submit"]' ).click();

    expect( onSubmitMock.mock.calls.length ).toBe( 1 );
    expect( onSubmitMock.mock.calls[ 0 ].length ).toBe( 1 );
    const data = onSubmitMock.mock.calls[ 0 ][ 0 ];
    expect( Object.keys( data ) ).toEqual( [ 'name' ] );
    expect( data.name ).toBe( formData.name );
} );

it( 'form validation (name)', () => {
    const onSubmitMock = jest.fn();
    const { widget } = renderer( ChainForm, {
        saveButtonText: 'Create',
        directives: {
            disabled,
            ...directives
        },
        onSubmit: onSubmitMock
    } );

    widget.querySelector( '[type="submit"]' ).click();
    expect( onSubmitMock.mock.calls.length ).toBe( 0 );

    widget.querySelector( '[type="submit"]' ).click();
    expect( onSubmitMock.mock.calls.length ).toBe( 0 );
} );
