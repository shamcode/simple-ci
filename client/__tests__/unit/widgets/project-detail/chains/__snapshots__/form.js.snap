// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "onsubmit": [Function],
    },
  },
  "html": "
<div class=\\"project-chain-form\\">
  <form>
    <div class=\\"group\\">
      <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Name</label>
    </div>
    <div class=\\"group\\">
      <button type=\\"submit\\"><span></span></button>
    </div>
  </form>
</div>
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "dataSaving": false,
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "onsubmit": [Function],
    },
    "name": "",
    "onSubmit": [Function],
    "renderAsync": false,
    "renderDependence": Array [],
    "saveButtonText": "",
    "types": Array [],
  },
}
`;
